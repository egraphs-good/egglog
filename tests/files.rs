use egg_smol::*;

fn test_program(program: &str, message: &str) {
    let mut egraph = EGraph::default();
    match egraph.parse_and_run_program(program) {
        Ok(msgs) => {
            for msg in msgs {
                log::info!("  {}", msg);
            }
        }
        Err(err) => panic!("{}: {err}", message),
    }
}

fn run(path: &str) {
    let _ = env_logger::builder().is_test(true).try_init();
    let program = std::fs::read_to_string(path).unwrap();
    test_program(&program, "Top level error");

    let egraph = EGraph::default();
    let program_str = egraph
        .parse_program(&program)
        .unwrap()
        .into_iter()
        .map(|x| x.to_string())
        .collect::<Vec<String>>()
        .join("\n");
    test_program(
        &program_str,
        &format!(
            "Program:\n{}\n ERROR after parse, to_string, and parse again.",
            program_str
        ),
    );
}

// include the tests generated from the build script
include!(concat!(std::env!("OUT_DIR"), "/files.rs"));
