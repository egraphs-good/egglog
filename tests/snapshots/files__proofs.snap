---
source: tests/files.rs
expression: serialized
---
nodes:
  Child1-2598150418935018079:
    op: Child1
    children: []
    eclass: "36"
    cost: 100000
  Child1-8469931425499020532:
    op: Child1
    children:
      - Child2-12055403783775160830
    eclass: "67"
    cost: 100000
  Child1-13770179520251441998:
    op: Child1
    children:
      - AstAdd-3683774522220528002
    eclass: "151"
    cost: 100000
  Child1-3741393243599718260:
    op: Child1
    children:
      - Child2-12003430155333968176
    eclass: "165"
    cost: 100000
  Child1-4312926155411299247:
    op: Child1
    children: []
    eclass: "164"
    cost: 100000
  Child1-9613174250163720713:
    op: Child1
    children:
      - AstAdd-16971738888365666495
    eclass: "66"
    cost: 100000
  Child1-5248274466311228812:
    op: Child1
    children:
      - AstAdd-16872052778055009793
    eclass: "91"
    cost: 100000
  Child1-16264382682304074769:
    op: Child1
    children:
      - AstAdd-1634549299562345488
    eclass: "140"
    cost: 100000
  Child1-1091269196223507527:
    op: Child1
    children:
      - AstAdd-13756935163900948186
    eclass: "106"
    cost: 100000
  Child1-1195216453105892835:
    op: Child1
    children:
      - Child1-13770179520251441998
    eclass: "152"
    cost: 100000
  add2x-0:
    op: add2x
    children: []
    eclass: "36"
    cost: 9223372036854776000
  PrfOf-16783941965674463102:
    op: PrfOf
    children:
      - VarRep-7017119389452091262
    eclass: "5"
    cost: 100000
  PrfOf-15952540911656918845:
    op: PrfOf
    children: []
    eclass: "8"
    cost: 100000
  PrfOf-2546176790493825425:
    op: PrfOf
    children: []
    eclass: "12"
    cost: 100000
  PrfOf-12626936695586741817:
    op: PrfOf
    children: []
    eclass: "20"
    cost: 100000
  PrfOf-10964134587551653303:
    op: PrfOf
    children: []
    eclass: "26"
    cost: 100000
  PrfOf-15173113486080567242:
    op: PrfOf
    children: []
    eclass: "33"
    cost: 100000
  PrfOf-16887889222556848410:
    op: PrfOf
    children: []
    eclass: "49"
    cost: 100000
  PrfOf-6859102945905124672:
    op: PrfOf
    children: []
    eclass: "63"
    cost: 100000
  PrfOf-8573878682381405840:
    op: PrfOf
    children: []
    eclass: "79"
    cost: 100000
  PrfOf-16991836479439233718:
    op: PrfOf
    children: []
    eclass: "93"
    cost: 100000
  Child2-8469931425499020532:
    op: Child2
    children:
      - Child1-2598150418935018079
    eclass: "66"
    cost: 100000
  Child2-13770179520251441998:
    op: Child2
    children:
      - AstAdd-3683774522220528002
    eclass: "152"
    cost: 100000
  Child2-9613174250163720713:
    op: Child2
    children:
      - Child1-3741393243599718260
    eclass: "67"
    cost: 100000
  Child2-5248274466311228812:
    op: Child2
    children:
      - AstAdd-16872052778055009793
    eclass: "67"
    cost: 100000
  Child2-12055403783775160830:
    op: Child2
    children: []
    eclass: "36"
    cost: 100000
  Child2-16264382682304074769:
    op: Child2
    children:
      - AstAdd-1634549299562345488
    eclass: "66"
    cost: 100000
  Child2-17823237533456777975:
    op: Child2
    children:
      - Child1-5248274466311228812
    eclass: "66"
    cost: 100000
  Child2-10392601675740072316:
    op: Child2
    children:
      - Child1-16264382682304074769
    eclass: "67"
    cost: 100000
  Child2-12003430155333968176:
    op: Child2
    children:
      - AstAdd-9680441442781277384
    eclass: "164"
    cost: 100000
  Child2-1091269196223507527:
    op: Child2
    children:
      - AstAdd-13756935163900948186
    eclass: "165"
    cost: 100000
  GetProof-155920885323577962:
    op: GetProof
    children:
      - Child1-9613174250163720713
    eclass: "5"
    cost: 1000000000
  GetProof-6027701891887580415:
    op: GetProof
    children:
      - Child2-9613174250163720713
    eclass: "8"
    cost: 1000000000
  GetProof-16264382682304074769:
    op: GetProof
    children:
      - AstAdd-1634549299562345488
    eclass: "142"
    cost: 1000000000
  GetProof-9613174250163720713:
    op: GetProof
    children:
      - AstAdd-16971738888365666495
    eclass: "12"
    cost: 1000000000
  GetProof-1195216453105892835:
    op: GetProof
    children:
      - AstAdd-15552222449345279837
    eclass: "79"
    cost: 1000000000
  GetProof-3741393243599718260:
    op: GetProof
    children:
      - Child2-12003430155333968176
    eclass: "20"
    cost: 1000000000
  GetProof-1091269196223507527:
    op: GetProof
    children:
      - AstAdd-13756935163900948186
    eclass: "108"
    cost: 1000000000
  GetProof-5248274466311228812:
    op: GetProof
    children:
      - AstAdd-16872052778055009793
    eclass: "93"
    cost: 1000000000
  GetProof-8469931425499020532:
    op: GetProof
    children:
      - Child2-12055403783775160830
    eclass: "26"
    cost: 1000000000
  addzero-0:
    op: addzero
    children: []
    eclass: "164"
    cost: 9223372036854776000
  AstAdd-7015864897269497032:
    op: AstAdd
    children:
      - Child2-5248274466311228812
      - Child2-8469931425499020532
    eclass: "36"
    cost: 1
  AstAdd-16872052778055009793:
    op: AstAdd
    children:
      - AstAdd-16175728989598053478
      - Child2-10392601675740072316
    eclass: "92"
    cost: 1
  AstAdd-1634549299562345488:
    op: AstAdd
    children:
      - Child1-16264382682304074769
      - Child2-16264382682304074769
    eclass: "141"
    cost: 1
  AstAdd-13756935163900948186:
    op: AstAdd
    children:
      - Child1-1091269196223507527
      - Child1-3741393243599718260
    eclass: "107"
    cost: 1
  AstAdd-16971738888365666495:
    op: AstAdd
    children:
      - Child2-17823237533456777975
      - Child1-8469931425499020532
    eclass: "165"
    cost: 1
  AstAdd-3683774522220528002:
    op: AstAdd
    children:
      - Child1-13770179520251441998
      - Child2-13770179520251441998
    eclass: "150"
    cost: 1
  AstAdd-9680441442781277384:
    op: AstAdd
    children:
      - Child1-1195216453105892835
      - Child1-4312926155411299247
    eclass: "112"
    cost: 1
  AstAdd-15552222449345279837:
    op: AstAdd
    children:
      - Child2-13770179520251441998
      - AstAdd-16971738888365666495
    eclass: "151"
    cost: 1
  AstAdd-16175728989598053478:
    op: AstAdd
    children:
      - Child1-1195216453105892835
      - AstVar-7017119389452091262
    eclass: "91"
    cost: 1
  AstAdd-7134264652287451959:
    op: AstAdd
    children:
      - Child2-13770179520251441998
      - Child1-13770179520251441998
    eclass: "99"
    cost: 1
  String-7017119389452091262:
    op: "\"x\""
    children: []
    eclass: String-7017119389452091262
    cost: 0
  AstVar-7017119389452091262:
    op: AstVar
    children:
      - String-7017119389452091262
    eclass: "66"
    cost: 1
  ProofBetween-14660223427748781285:
    op: ProofBetween
    children:
      - Child2-8469931425499020532
      - Child2-16264382682304074769
      - Child2-17823237533456777975
    eclass: "0"
    cost: 100000
  ProofBetween-18406589706221332193:
    op: ProofBetween
    children:
      - Child1-1195216453105892835
      - Child2-13770179520251441998
      - Child1-1195216453105892835
    eclass: "0"
    cost: 100000
  ProofBetween-16555661497296228401:
    op: ProofBetween
    children:
      - Child1-3741393243599718260
      - Child2-12003430155333968176
      - Child2-1091269196223507527
    eclass: "0"
    cost: 100000
  ProofBetween-4520260267243263193:
    op: ProofBetween
    children:
      - AstAdd-16971738888365666495
      - Child1-2598150418935018079
      - Child1-3741393243599718260
    eclass: "0"
    cost: 100000
  ProofBetween-2872005694695115899:
    op: ProofBetween
    children:
      - Child2-1091269196223507527
      - AstAdd-15552222449345279837
      - AstAdd-16971738888365666495
    eclass: "0"
    cost: 100000
  ProofBetween-6935811013450972484:
    op: ProofBetween
    children:
      - Child1-3741393243599718260
      - AstAdd-16872052778055009793
      - Child2-1091269196223507527
    eclass: "0"
    cost: 100000
  ProofBetween-8816407285316904857:
    op: ProofBetween
    children:
      - Child2-9613174250163720713
      - Child2-5248274466311228812
      - Child2-10392601675740072316
    eclass: "0"
    cost: 100000
  ProofBetween-1195769155606790371:
    op: ProofBetween
    children:
      - AstAdd-16971738888365666495
      - Child1-3741393243599718260
      - Child2-1091269196223507527
    eclass: "0"
    cost: 100000
  ProofBetween-13770732222752339534:
    op: ProofBetween
    children:
      - AstAdd-16971738888365666495
      - Child1-3741393243599718260
      - addzero-0
    eclass: "177"
    cost: 100000
  ProofBetween-52526330942090190:
    op: ProofBetween
    children:
      - Child2-1091269196223507527
      - AstAdd-16971738888365666495
      - add2x-0
    eclass: "179"
    cost: 100000
  VarRep-7017119389452091262:
    op: VarRep
    children:
      - String-7017119389452091262
    eclass: "6"
    cost: 1
root_eclasses: []
class_data:
  "36":
    type: AstMath
  "35":
    type: AstMath
  "67":
    type: AstMath
  "151":
    type: AstMath
  "150":
    type: AstMath
  "165":
    type: AstMath
  "164":
    type: AstMath
  "51":
    type: AstMath
  "66":
    type: AstMath
  "91":
    type: AstMath
  "92":
    type: AstMath
  "140":
    type: AstMath
  "141":
    type: AstMath
  "106":
    type: AstMath
  "107":
    type: AstMath
  "152":
    type: AstMath
  "5":
    type: Proof
  "6":
    type: TrmPrf
  "8":
    type: Proof
  "9":
    type: TrmPrf
  "12":
    type: Proof
  "13":
    type: TrmPrf
  "20":
    type: Proof
  "21":
    type: TrmPrf
  "26":
    type: Proof
  "27":
    type: TrmPrf
  "33":
    type: Proof
  "34":
    type: TrmPrf
  "49":
    type: Proof
  "50":
    type: TrmPrf
  "63":
    type: Proof
  "64":
    type: TrmPrf
  "79":
    type: Proof
  "80":
    type: TrmPrf
  "93":
    type: Proof
  "94":
    type: TrmPrf
  "134":
    type: AstMath
  "112":
    type: AstMath
  "142":
    type: Proof
  "108":
    type: Proof
  "99":
    type: AstMath
  String-7017119389452091262:
    type: String
  "0":
    type: ProofList
  "177":
    type: ProofList
  "179":
    type: ProofList

